pyro/infer/mcmc/adaptation.py

```diff
@@ -194,25 +176,177 @@ def step(self, t, z, accept_prob):
+
     @property
     def inverse_mass_matrix(self):
         return self._inverse_mass_matrix
 
     @inverse_mass_matrix.setter
     def inverse_mass_matrix(self, value):
-        # XXX: for backward compatibility (when users want to specify a fixed inverse mass matrix)
-        if torch.is_tensor(value):
-            assert len(self._inverse_mass_matrix) == 1
-            all_sites = list(self._inverse_mass_matrix.keys())[0]
-            value = {all_sites: value}
-
-        for site_names, inv_mass_matrix in value.items():
-            self._inverse_mass_matrix[site_names] = inv_mass_matrix
-        self._update_r_dist()
-        if self.adapt_mass_matrix:
-            for site_names, adapt_scheme in self._mass_matrix_adapt_scheme.items():
-                adapt_scheme.reset()
+        for site_names, inverse_mass_matrix in value.items():
+            if site_names in self._adapt_scheme:
+                self._adapt_scheme[site_names].reset()
+            mass_matrix_sqrt_inverse = _transpose(_cholesky(inverse_mass_matrix))
+            mass_matrix_sqrt = _upper_triangular_inverse(mass_matrix_sqrt_inverse)
+            self._inverse_mass_matrix[site_names] = inverse_mass_matrix
+            self._mass_matrix_sqrt[site_names] = mass_matrix_sqrt
+            self._mass_matrix_sqrt_inverse[site_names] = mass_matrix_sqrt_inverse
 
-    @property
-    def r_dist(self):
-        return self._r_dist
+    def configure(self, mass_matrix_shape, adapt_mass_matrix=True, options={}):
+        """
+        Setups initial mass matrix.
+
+        :param dict mass_matrix_shape: a dict that maps tuples of site names to the shape of
+            the corresponding mass matrix. Each tuple of site names corresponds to a block.
+        :param bool adapt_mass_matrix: a flag to decide whether an adaptation scheme will be used.
+        :param dict options: tensor options to construct the initial mass matrix.
+        """
+        inverse_mass_matrix = {}
+        for site_names, shape in mass_matrix_shape.items():
+            self._mass_matrix_size[site_names] = shape[0]
+            diagonal = len(shape) == 1
+            inverse_mass_matrix[site_names] = torch.full(shape, self._init_scale, **options) \
+                if diagonal else torch.eye(*shape, **options) * self._init_scale
+            if adapt_mass_matrix:
+                adapt_scheme = WelfordCovariance(diagonal=diagonal)
+                self._adapt_scheme[site_names] = adapt_scheme
+
+        self.inverse_mass_matrix = inverse_mass_matrix
+
+    def update(self, z, z_grad):
+        """
+        Updates the adaptation scheme using the new sample `z` or its grad `z_grad`.
+
+        :param dict z: the current value.
+        :param dict z_grad: grad of the current value.
+        """
+        for site_names, adapt_scheme in self._adapt_scheme.items():
+            z_flat = torch.cat([z[name].reshape(-1) for name in site_names])
+            adapt_scheme.update(z_flat.detach())
```

Comment:
nit: Maybe detach earlier `z[name].detach()` (or even earlier if possible)?

Reply:
yes, I think it is better.